src/pymoca/tree.py

```diff
@@ -889,13 +894,38 @@ def _instantiate_class(
         # 2.2 Copy local contents into the element itself
         self._copy_class_contents(new_class, copy_extends=True, partially=partially)
 
+        # Make a copy of the extends clauses for useful error messages later
+        extends_clauses = new_class.extends.copy()
+
         # 3. Instantiate extends and 4. Check extends class lookup
         for index, extends in enumerate(new_class.extends):
             extends_instance = self._instantiate_extends(
                 extends, modification_environment, new_class, partially
             )
             new_class.extends[index] = extends_instance
 
+        # Check we do not extend from any symbols/classes inherited
+        extends_names = {
+            _parse_str_or_ref(e.component)[0]: str(e.component) for e in extends_clauses
+        }
+
+        for extends_name, extends_component_ref in extends_names.items():
+            if extends_name in self.BUILTIN_TYPES:
+                # Built-in classes contain a symbol with the same name (which
+                # would cause an error). Note that this is an implementation
+                # detail where we differ a little from the spec at the moment.
+                continue
+            for other_class in new_class.extends:
+                other_names = {
+                    *other_class.ast_ref.symbols.keys(),
+                    *other_class.ast_ref.classes.keys(),
+                }
+                if extends_name in other_names:
+                    raise ModelicaSemanticError(
+                        f"Cannot extend '{new_class.full_reference()}' with '{extends_component_ref}'; "
+                        f"'{extends_name}' also exists in names inherited from '{other_class.ast_ref.name}'"
+                    )
+
```

Comment:
You said you don't like that the cannot extend with and inherited name check is done here rather than in `_instantiate_extends`. I agree. A simple remedy is to move the whole `for index, extends in enumerate(new_class.extends):` block as well as the error check code into `_instantiate_extends`. This makes it a little less symmetric with the previous code looping on local classes and symbols, but extends are a little different anyway.

Reply:
I can try that again, but the issue I run into is that I need both the original extends clause, and also all classes to have been looked up. And _inside_ `_instantiate_extends` I cannot do that, some `parent.extends` entries are of type `InstanceExtends`, some are `ExtendsClause`. I need the `ExtendsClause` to know how the user referred to the type for a proper error, and I need the `InstanceExtends` to know the symbols/classes.

Reply:
> One way to fix this would be to make _instantiate_extends itself do the loop on all of the extends clauses

This triggered me to make it like in 0f75df9611754a5bac6092366e6285d040b35fad (pushed it on top, will squash if OK by you).

That probably also makes sense with the check of the spec (still a TODO in your code). I don't think this is easily done in `_instantiate_extends_single`, as the spec says it needs to be done after handling _all_ of them:

> The classes of extends-clauses are looked up before and after handling extends-clauses; and it is an error if those lookups generate different results.

Reply:
Agreed. Good change and good for future changes on top. Squash-away! I'm not doing anything currently.